src/ALCProxy.Communication/ICommObjects.cs

```diff
@@ -0,0 +1,257 @@
+        /// <returns></returns>
+        public object SendMethod(MethodInfo method, object[] args)
+        {
+            if (_server == null) //We've called the ALC unload, so the proxy has been cut off
+            {
+                throw new InvalidOperationException("Error in ALCClient: Proxy has been unloaded, or communication server was never set up correctly");
+            }
+            SerializeParameters(args, out IList<object> streams, out IList<Type> argTypes);
+            object encryptedReturn = _callMethod.Invoke(_server, new object[] { method, streams, argTypes });
+            return DecryptReturnType(encryptedReturn, method.ReturnType);
+        }
+        protected void SerializeParameters(object[] arguments, out IList<object> serializedArgs, out IList<Type> argTypes)
+        {
+            argTypes = new List<Type>();
+            serializedArgs = new List<object>();
+            for (int i = 0; i < arguments.Length; i++)
+            {
+                object arg = arguments[i];
+                Type t = arg.GetType();
+                //Serialize the argument
+                object serialArg = SerializeParameter(arg, t);
+                serializedArgs.Add(serialArg);
+                argTypes.Add(t);
+            }
+        }
+        protected abstract object SerializeParameter(object param, Type paramType);
+        protected abstract object DecryptReturnType(object returnedObject, Type returnType);
+    }
+    public abstract class ALCServer<ObjectInterface> : IProxyServer
+    {
+        public object instance;
+        public AssemblyLoadContext currentLoadContext;
+        public ALCServer(Type instanceType, Type[] genericTypes, IList<object> serializedConstParams, IList<Type> constArgTypes)
+        {
+            currentLoadContext = AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
+            if (genericTypes != null)
+            {
+                instanceType = instanceType.MakeGenericType(genericTypes.Select(x => ConvertType(x)).ToArray());
+            }
+            var constructorParams = DecryptParameters(serializedConstParams, constArgTypes);
+            SetInstance(instanceType, constArgTypes.ToArray(), constructorParams);
+        }
+        /// <summary>
+        /// Create the instance of the object we want to proxy
+        /// </summary>
+        /// <param name="instanceType">the type of the object we want</param>
+        /// <param name="constructorTypes">The list of types that the constructor of the object takes in as an argument</param>
+        /// <param name="constructorArgs">The physical objects that are the parameters to the constructor</param>
+        protected void SetInstance(Type instanceType, Type[] constructorTypes, object[] constructorArgs)
+        {
+            ConstructorInfo ci = instanceType.GetConstructor(constructorTypes);//GrabConstructor(constructorTypes, instanceType);//instanceType.GetConstructor(constructorTypes);
```

Comment:
nit: feel free to use var when the type is obvious from the right hand expression